File: packages/totem-sdk/packages/manifest/rust/src/verify.rs
use crate::canonical::manifest_digest;
use crate::types::{SignedManifest, VerifyResult};

fn manifest_address_field(manifest: &serde_json::Value) -> String {
    match manifest["type"].as_str().unwrap_or("") {
        "app" => manifest["authorAddress"].as_str().unwrap_or("").to_string(),
        "capability" => manifest["agentAddress"].as_str().unwrap_or("").to_string(),
        "dapp" => manifest["authorAddress"].as_str().unwrap_or("").to_string(),
        "edge-service" => manifest["operatorAddress"].as_str().unwrap_or("").to_string(),
        _ => String::new(),
    }
}

pub fn verify_manifest(
    signed: &SignedManifest,
    verify_sig_fn: &dyn Fn(&str, &[u8], &str) -> bool,
) -> VerifyResult {
    let digest = manifest_digest(&signed.manifest);

    if !verify_sig_fn(&signed.signature, &digest, &signed.signer_public_key) {
        return VerifyResult {
            valid: false,
            reason: Some("WOTS signature invalid".to_string()),
            signer_address: signed.author_address.clone(),
        };
    }

    let expected_address = manifest_address_field(&signed.manifest);
    if signed.author_address != expected_address {
        return VerifyResult {
            valid: false,
            reason: Some(format!(
                "authorAddress mismatch: signed by '{}' but manifest declares '{}'",
                signed.author_address, expected_address
            )),
            signer_address: signed.author_address.clone(),
        };
    }

    VerifyResult {
        valid: true,
        reason: None,
        signer_address: signed.author_address.clone(),
    }
}

File: packages/manifest/src/verify.ts
/**
 * verifyManifest — verifies a SignedManifest without external state.
 *
 * Steps:
 *   1. Recompute the canonical manifest digest.
 *   2. Verify the WOTS signature against the stored 32-byte PKdigest
 *      (signerPublicKey field) using wotsVerifyDigest.
 *   3. Confirm authorAddress matches the manifest's address field.
 *
 * wotsVerifyDigest is used (rather than wotsVerify) because signerPublicKey
 * stores the 32-byte WOTS PKdigest as returned by wotsKeypairFromSeed.kp.pk.
 * The full 1088-byte key is not stored in SignedManifest to keep it compact.
 */

import { wotsVerifyDigest, hexToBytes } from '@totemsdk/core';
import type { SignedManifest, VerifyResult } from './types.js';
import { manifestDigest } from './sign.js';

function manifestAddressField(manifest: SignedManifest['manifest']): string {
  switch (manifest.type) {
    case 'app':          return manifest.authorAddress;
    case 'capability':   return manifest.agentAddress;
    case 'dapp':         return manifest.authorAddress;
    case 'edge-service': return manifest.operatorAddress;
    default: {
      const _exhaustive: never = manifest;
      throw new Error(`verifyManifest: unknown manifest type: ${JSON.stringify(_exhaustive)}`);
    }
  }
}

export function verifyManifest(signed: SignedManifest): VerifyResult {
  const { manifest, signature, signerPublicKey, authorAddress } = signed;

  let sigBytes: Uint8Array;
  let pkDigest: Uint8Array;
  try {
    sigBytes = hexToBytes(signature);
    pkDigest = hexToBytes(signerPublicKey);
  } catch (e) {
    return { valid: false, reason: `hex decode failed: ${String(e)}`, signerAddress: authorAddress };
  }

  const digest = manifestDigest(manifest);

  let sigValid: boolean;
  try {
    sigValid = wotsVerifyDigest(sigBytes, digest, pkDigest);
  } catch (e) {
    return { valid: false, reason: `WOTS verify threw: ${String(e)}`, signerAddress: authorAddress };
  }

  if (!sigValid) {
    return { valid: false, reason: 'WOTS signature invalid', signerAddress: authorAddress };
  }

  const expectedAddress = manifestAddressField(manifest);
  if (authorAddress !== expectedAddress) {
    return {
      valid: false,
      reason: `authorAddress mismatch: signed by '${authorAddress}' but manifest declares '${expectedAddress}'`,
      signerAddress: authorAddress,
    };
  }

  return { valid: true, signerAddress: authorAddress };
}
